### PR TITLE
`General`: Add course exercise overview grouped by due date (timeframe view)

### PR DESCRIPTION
**Pull request by florian-glombik**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://ls1intum.github.io/Artemis/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://ls1intum.github.io/Artemis/dev/development-process/#naming-conventions-for-github-pull-requests).

#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
Calculations in parent component for the weekly view will be moved to the child component in a follow up
- [x] I followed the [coding and design guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client/).
- [x] Following the [theming guidelines](https://ls1intum.github.io/Artemis/dev/guidelines/client-design/), I specified colors only in the theming variable files and checked that the changes look consistent in both the light and the dark theme.
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

Tests will be added in a Follow Up.

### Description

Improves the UX of the exercise Overview

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Student
- 1 Course with / without exercises

1. View the `Course Overview` (see the exercise List)
2. See that the exercises are ordered properly within the groups (by due date, if equal by name)
3. Test that exercises can be expanded / collapsed 
4. Verify that filters work
5. Verify that search works _(that the categories do not uncollapse is a known UX issue)_
6. Verify that 1 Category is uncollapsed in the `Timeframe view` but `Past` shall never be uncollapsed by default when loading the page

~~Something I did not check yet: the guided tour may be broken (open TODO)~~
Checked on ts1, no guided tour seems to exist there

#### Performance Review
- [ ] I (as a reviewer) confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 

#### Code Review
- [x] Code Review 1
- [ ] Code Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage
Will be fixed and added in a follow up / before merge if time left

### Screenshots

NEW Version
image

old Version
image

### CODE REVIEW

File: src/main/webapp/i18n/en/student-dashboard.json [hunk 1]
```diff
@@ -56,12 +56,18 @@
                 "sortExercises": "Sort exercises",
                 "newFirst": "Newest",
                 "oldFirst": "Oldest",
+                "previous": "Previous",
+                "current": "Current",
+                "future": "Future",
+                "noDueDate": "No due date",
+                "byDueDate": "by due date",
```
Comment (JohannesStoehr) [inline src/main/webapp/i18n/en/student-dashboard.json:63]: I would assume that "by due date" means that exercises with the same due date are grouped together. I'm not sure what a better wording would be...

**Review by JohannesStoehr — CHANGES_REQUESTED** [review]

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]
I don't quite understand how the exercises are ordered within their major grouping? As a student I would find it difficult to find an exercise if they are unsorted.

File: src/main/webapp/i18n/en/student-dashboard.json (hunk 1 above)
Comment (florian-glombik) [reply src/main/webapp/i18n/en/student-dashboard.json:63]: May "by urgency"?

**Comment by florian-glombik** [comment]
https://github.com/ls1intum/Artemis/pull/7363#pullrequestreview-1676619203
They are sorted by date and if the date is the same by name of the exercise - right now I am using the exact date for the date comparison

Which comparison would you prefer to start sorting by name?
1. Same Day?
2. Same Day + Same hour? 
3. Same Day + Within x hours?

**Comment by MaximilianAnzinger** [comment]
@florian-glombik now I get it. It's the release date no the due date.

File: src/main/webapp/i18n/de/student-dashboard.json [hunk 2]
```diff
@@ -56,12 +56,18 @@
                 "sortExercises": "Aufgaben sortieren",
                 "newFirst": "Neueste",
                 "oldFirst": "Älteste",
+                "previous": "Vorangegangen",
+                "current": "Aktuell",
+                "future": "Zukünftig",
+                "noDueDate": "Keine Einreichungsfrist",
```
Comment (MaximilianAnzinger) [inline src/main/webapp/i18n/de/student-dashboard.json:62]: `Unbefristet`?

**Review by MaximilianAnzinger — COMMENTED** [review]
Besides the suggestions by Johannes, I added one more. Other than that code looks good 👍

File: src/main/webapp/i18n/en/student-dashboard.json (hunk 1 above)
Comment (florian-glombik) [reply src/main/webapp/i18n/en/student-dashboard.json]: I have added new translations within https://github.com/ls1intum/Artemis/pull/7363/commits/ba777891b491f980343b929869a3c670aa9cab76, do you think they are better?

Comment (JohannesStoehr) [reply src/main/webapp/i18n/en/student-dashboard.json]: I like it

**Review by JohannesStoehr — DISMISSED** [review]
Code looks good to me now.
Some details can be handled in a follow-up like dynamically making „current“ bigger when there are no exercises due in the next seven days

**Review by MaximilianAnzinger — DISMISSED** [review]
Reapproved Code
Also tested manually on TS2 👍

File: src/main/webapp/app/overview/course-exercises/course-exercises-grouped-by-category.component.ts [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+
+        const dueDateIsWithinNextWeek = dueDate.isBefore(now.add(1, 'week'));
+        if (dueDateIsWithinNextWeek) {
+            return 'current';
+        }
+
+        return 'future';
+    }
+
+    protected readonly Object = Object;
```
Comment (lennart-keller) [inline src/main/webapp/app/overview/course-exercises/course-exercises-grouped-by-category.component.ts:146]: Usually, these are defined at the top of the class

**Review by lennart-keller — APPROVED** [review]
Code lgtm, besides one minor comment. Considering the urgency of this PR, this can be fixed in a follow up, when tests are added. 

Tested on TS2. Everything works as described in the testing steps.

**Review by basak-akan — DISMISSED** [review]
Tested on TS1, and worked as described. 
- Exercises are ordered correctly within the groups and can be expanded/collapsed.
- Search and filters worked well

File: src/main/webapp/app/overview/course-exercises/course-exercises-grouped-by-category.component.ts (hunk 3 above)
Comment (florian-glombik) [reply src/main/webapp/app/overview/course-exercises/course-exercises-grouped-by-category.component.ts]: Some changes were needed anyways - I've moved it to the top now

**Review by JohannesStoehr — DISMISSED** [review]
Approve code after latest changes

**Review by laurenzfb — APPROVED** [review]
Tested in testing session. LGTM

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session on ts1. New overview looks good. Nice improvement

**Review by max-bergmann — APPROVED** [review]
Tested in testing session. Nice improvement

**Review by JohannesStoehr — APPROVED** [review]
Code

**Pull request merged**